llvm/lib/CodeGen/MachineBlockPlacement.cpp

```diff
@@ -2614,18 +2614,21 @@ MachineBlockPlacement::collectLoopBlockSet(const MachineLoop &L) {
         LoopFreq += MBFI->getBlockFreq(LoopPred) *
                     MBPI->getEdgeProbability(LoopPred, L.getHeader());
 
-    for (MachineBasicBlock *LoopBB : L.getBlocks()) {
-      if (LoopBlockSet.count(LoopBB))
+    for (auto &MBB : *F) {
```

Comment:
IIUC, the original patch avoids traversing the full function body to avoid the high complexity in compile time. This will probably restore the old behavior.